File: .apps/ui/src/types/events.ts
/**
 * TypeScript type definitions for events API
 */

/**
 * Event data structure from the API
 */
export interface EventData {
  event_id: string;
  aggregate_id: string;
  aggregate_type: string;
  event_type: string;
  aggregate_version: number;
  timestamp: string;
  user_id?: string;
  causation_id?: string;
  correlation_id?: string;
  name?: string;
  attributes?: Record<string, unknown>;
  trace_id?: string;
  span_id?: string;
  kind?: string;
  status?: {
    code: string;
  };
  
  // Agent-specific fields
  agent_name?: string;
  agent_id?: string;
  parent_agent_id?: string;
  workflow_id?: string;
  event_name?: string;
  
  // Workflow-specific fields
  user_prompt?: string;
  
  // System-specific fields
  session_id?: string;
  
  // Computed/helper field for UI components
  aggregate?: string;
  
  // Additional UI fields
  id?: string;
  eventName?: string;
}

/**
 * Server-Sent Event message structure
 */
export interface SSEMessage {
  event: 'event_created' | 'heartbeat' | 'error';
  data: string;
  id?: string;
}

/**
 * Connection status for SSE
 */
export interface ConnectionStatus {
  connected: boolean;
  lastHeartbeat?: Date;
  error?: string;
}

/**
 * Event filters for querying
 */
export interface EventFilters {
  event_type?: string;
  limit?: number;
  offset?: number;
}

File: .apps/listener/models/events.py
"""Event models for the listener."""

from datetime import datetime
from enum import Enum
from typing import Dict, Any, Optional
from pydantic import BaseModel, Field


class EventType(str, Enum):
    """Supported event types."""
    AGENT_EVENT = "agent_event"
    WORKFLOW_EVENT = "workflow_event"
    SYSTEM_EVENT = "system_event"


class AggregateType(str, Enum):
    """Supported aggregate types."""
    AGENT_AGGREGATE = "agent_aggregate"
    WORKFLOW_AGGREGATE = "workflow_aggregate"
    SYSTEM_AGGREGATE = "system_aggregate"


class StreamEventType(str, Enum):
    """Types of events from the stream."""
    EVENT_CREATED = "event_created"
    HEARTBEAT = "heartbeat"
    ERROR = "error"


class EventData(BaseModel):
    """Represents event data from the stream."""
    
    event_id: Optional[str] = None
    event_type: Optional[str] = None
    event_name: Optional[str] = None
    aggregate_type: Optional[str] = None
    aggregate_id: Optional[str] = None
    correlation_id: Optional[str] = None
    causation_id: Optional[str] = None
    attributes: Dict[str, Any] = Field(default_factory=dict)
    timestamp: Optional[datetime] = None
    
    class Config:
        """Pydantic configuration."""
        extra = "allow"  # Allow additional fields


class StreamItem(BaseModel):
    """Represents an item from the event stream."""
    
    event: str
    data: Optional[Dict[str, Any]] = None
    
    class Config:
        """Pydantic configuration."""
        extra = "allow"